Keep raw settings on missing variable

MickaJsd opened the pull request:
Add parameter keepNonResolvedVariables to ConfigurationSubstitutor
Update ApplySubstitution : on non resolved variable without fallback value, if configured for not throwing on missing variable, keep original setting.

Use case: I want to access my invalid settings to log them and manage the exception throwing by myself

MickaJsd committed:
Keep raw settings on missing variable

Add parameter keepNonResolvedVariables to ConfigurationSubstitutor
Update ApplySubstitution : on non resolved variable without fallback value, if configured for not throwing on missing variable, keep original setting

molinch commented on ConfigurationSubstitutor/ConfigurationSubstitutor.cs:
```diff
@@ -94,12 +96,18 @@ private string ApplySubstitution(IConfiguration configuration, string value, Has
                     throw new UndefinedConfigVariableException($"{_startsWith}{capture.Value}{_endsWith}");
                 }
+
+                if (substitutedValue == null && _keepNonResolvedVariables)
```
Might be better to introduce a breaking change, we start to have to have too many flags.
Instead of having both exceptionOnMissingVariables and keepNonResolvedVariables, I would rather something like:
```c#
public enum UnresolvedVariableBehaviour {
    // UndefinedConfigVariableException is thrown for unresolved variables
    Throw,
    // Unresolved variables are ignored and replaced by empty content
    IgnorePattern,
    // The pattern for substitution is kept in case of unresolved variables 
    KeepPattern
}
```

Then the default behaviour would remain to throw and there's the possibility to Ignore (currently that's exceptionOnMissingVariables:false, or keep the variable pattern)

MickaJsd replied:
I added new signatures and tagged old method as obsolete.
I kept all previous unit test to guarantee that deprecated methods are still OK.

molinch replied:
Thanks @MickaJsd we are almost there, good job!

MickaJsd committed:
Use enum to set unresolved variables behavior

Create "UnresolvedVariableBehaviour" enum to manage unresolved variables behavior in "ApplySubstitution" method

To avoid breaking change on next release :
Add "Obsolete" tag on methods setting unresolved variables behavior with exceptionOnMissingVariables boolean
Add unit test for deprecated methods to keep them until deprecated methods are removed

MickaJsd commented on ConfigurationSubstitution.Tests/ConfigurationTests.cs:
```diff
@@ -1,19 +1,763 @@
+                () => new ConfigurationManager()
+            };
+
+        #region Obsolete method tests
```
This region contains unit test on deprecated methods

MickaJsd commented on ConfigurationSubstitutor/IConfigurationBuilderExtensions.cs:
```diff
@@ -1,27 +1,54 @@
+
+        #endregion
+
+        #region Private Methods
```
Put the private method aside from the public ones

molinch replied:
Fine for me but can we remove regions?

molinch commented on ConfigurationSubstitution.Tests/ConfigurationTests.cs:
```diff
@@ -746,8 +1484,46 @@ public void Should_get_fallback_default_values_substituted_when_multiple_matches
+
+        [Theory]
+        [MemberData(nameof(ConfigurationBuilderTheoryData))]
+        public void Should_get_unsubstituted_value_when_keepNonResolvedVariables_is_true(Func<IConfigurationBuilder> builderGenerator)
```
Method name can be changed now

MickaJsd replied:
My bad

molinch commented:
Thanks @MickaJsd I will publish a Nuget package soon

molinch commented:
1.5 is on nuget now https://www.nuget.org/packages/ConfigurationSubstitutor/
